code/modules/mob/living/basic/lavaland/watcher/watcher_overwatch.dm

```diff
@@ -0,0 +1,158 @@
+	StartCooldown()
+	return TRUE
+
+/// Status effect which tracks whether our overwatched mob moves or acts
+/datum/status_effect/overwatch
+	id = "watcher_overwatch"
+	duration = 5 SECONDS
+	status_type = STATUS_EFFECT_MULTIPLE
+	alert_type = /atom/movable/screen/alert/status_effect/overwatch
+	/// Distance at which we break off the ability
+	var/watch_range = 9
+	/// Visual effect to make the status obvious
+	var/datum/beam/link
+	/// Which watcher is watching?
+	var/mob/living/watcher
+	/// Type of projectile to fire
+	var/projectile_type
+	/// Noise to make when we shoot beam
+	var/projectile_sound
+	/// Did the overwatch ever trigger during our run?
+	var/overwatch_triggered = FALSE
+	/// Signals which trigger a hostile response
+	var/static/list/forbidden_actions = list(
+		COMSIG_MOB_ABILITY_FINISHED,
+		COMSIG_MOB_ATTACK_HAND,
+		COMSIG_MOB_DROVE_MECH,
+		COMSIG_MOB_FIRED_GUN,
+		COMSIG_MOB_ITEM_ATTACK,
+		COMSIG_MOB_THROW,
+		COMSIG_MOB_USED_MECH_EQUIPMENT,
+		COMSIG_MOB_USED_MECH_MELEE,
+		COMSIG_MOVABLE_MOVED,
+	)
+
+/datum/status_effect/overwatch/on_creation(mob/living/new_owner, set_duration, mob/living/watcher, projectile_type, projectile_sound)
+	if (isnull(watcher) || isnull(projectile_type))
+		return FALSE
+	if (HAS_TRAIT(new_owner, TRAIT_OVERWATCH_IMMUNE))
+		return FALSE
+	src.watcher = watcher
+	src.projectile_type = projectile_type
+	src.projectile_sound = projectile_sound
+	if (!isnull(set_duration))
+		duration = set_duration
+	return ..()
+
+/datum/status_effect/overwatch/on_apply()
+	. = ..()
+	if (!.)
+		return FALSE
+	ADD_TRAIT(owner, TRAIT_OVERWATCHED, TRAIT_STATUS_EFFECT(id))
+	owner.do_alert_animation()
+	owner.Immobilize(0.25 SECONDS) // Just long enough that they don't trigger it by mistake
+	owner.playsound_local(owner, 'sound/machines/chime.ogg', 50, TRUE)
+	var/atom/beam_origin = ismecha(owner.loc) ? owner.loc : owner
+	link = beam_origin.Beam(watcher, icon_state = "r_beam", override_target_pixel_x = 0)
+	RegisterSignals(owner, forbidden_actions, PROC_REF(opportunity_attack))
+	RegisterSignals(owner, list(COMSIG_QDELETING, COMSIG_LIVING_DEATH), PROC_REF(on_participant_died))
+	RegisterSignals(watcher, list(COMSIG_QDELETING, COMSIG_LIVING_DEATH), PROC_REF(on_participant_died))
+
+/datum/status_effect/overwatch/on_remove()
+	UnregisterSignal(owner, forbidden_actions + list(COMSIG_QDELETING, COMSIG_LIVING_DEATH))
+	UnregisterSignal(owner, forbidden_actions)
+	QDEL_NULL(link)
+	REMOVE_TRAIT(owner, TRAIT_OVERWATCHED, TRAIT_STATUS_EFFECT(id))
+	if (!QDELETED(owner))
+		owner.apply_status_effect(/datum/status_effect/overwatch_immune)
+	return ..()
+
+/datum/status_effect/overwatch/Destroy()
+	QDEL_NULL(link)
+	if (isnull(watcher))
+		return ..()
+	if (!overwatch_triggered) // Side effects in Destroy? Well it turns out `on_remove` is also just called on Destroy. But only if the owner isn't deleting.
+		watcher.Stun(2 SECONDS, ignore_canstun = TRUE)
```

Comment:
can we at the very least `INVOKE_ASYNC` this so we can get out of dodge/hold up the `Destroy()` chain